File: cpp/02_medium/src/Food.cpp
#include "../include/Food.h"
#include <random>

Food::Food() : type(FoodType::NORMAL), points(10), 
               color(sf::Color::Red), lifetime(0), age(0) {
    respawn({});
}

void Food::respawn(const std::vector<sf::Vector2i>& occupiedPositions) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(0, 19);
    
    do {
        position.x = dis(gen);
        position.y = dis(gen);
    } while (std::find(occupiedPositions.begin(), occupiedPositions.end(), 
                       position) != occupiedPositions.end());
    
    // Случайный тип еды
    static std::uniform_real_distribution<> chance(0.0, 1.0);
    float r = chance(gen);
    
    if (r < 0.7f) {
        type = FoodType::NORMAL;
        points = 10;
        color = sf::Color::Red;
        lifetime = 0;
    } else if (r < 0.85f) {
        type = FoodType::BONUS;
        points = 50;
        color = sf::Color::Yellow;
        lifetime = 300; // 5 секунд при 60 FPS
    } else if (r < 0.93f) {
        type = FoodType::SPEED;
        points = 20;
        color = sf::Color::Cyan;
        lifetime = 0;
    } else if (r < 0.98f) {
        type = FoodType::SLOW;
        points = 20;
        color = sf::Color::Magenta;
        lifetime = 0;
    } else {
        type = FoodType::GOLDEN;
        points = 100;
        color = sf::Color(255, 215, 0);
        lifetime = 0;
    }
    
    age = 0;
}

void Food::update() {
    if (lifetime > 0) {
        age++;
    }
}

bool Food::isExpired() const {
    return lifetime > 0 && age >= lifetime;
}

void Food::draw(sf::RenderWindow& window, int cellSize) const {
    sf::CircleShape shape(cellSize / 2 - 2);
    shape.setPosition(position.x * cellSize + 2, position.y * cellSize + 2);
    
    // Пульсация для бонусной еды
    if (type == FoodType::BONUS) {
        float pulse = (sinf(age * 0.1f) + 1.0f) * 0.5f;
        sf::Color c = color;
        c.a = 255 * pulse;
        shape.setFillColor(c);
    } else {
        shape.setFillColor(color);
    }
    
    window.draw(shape);
    
    // Символ для особой еды
    sf::Font font;
    font.loadFromFile("resources/fonts/arial.ttf");
    
    sf::Text text;
    text.setFont(font);
    text.setCharacterSize(14);
    text.setFillColor(sf::Color::White);
    
    switch (type) {
        case FoodType::BONUS:
            text.setString("★");
            break;
        case FoodType::SPEED:
            text.setString("⚡");
            break;
        case FoodType::SLOW:
            text.setString("🐢");
            break;
        case FoodType::GOLDEN:
            text.setString("👑");
            break;
        default:
            return;
    }
    
    text.setPosition(position.x * cellSize + cellSize / 3,
                     position.y * cellSize + cellSize / 4);
    window.draw(text);
}

File: csharp/02_medium/Food.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace SnakeGameWinForms
{
    public enum FoodType
    {
        Normal,
        Bonus,
        Speed,
        Slow,
        Golden
    }

    public class Food
    {
        private static readonly Random random = new Random();
        private const int GridSize = 20;

        public Point Position { get; private set; }
        public FoodType Type { get; private set; }
        public int Points { get; private set; }
        public Color Color { get; private set; }

        public Food()
        {
            Type = FoodType.Normal;
            Points = 10;
            Color = Color.Red;
        }

        public void Respawn(List<Point> occupiedPositions)
        {
            int maxAttempts = 1000;
            bool validPosition = false;

            do
            {
                Position = new Point(random.Next(GridSize), random.Next(GridSize));
                maxAttempts--;

                if (!occupiedPositions.Contains(Position))
                {
                    validPosition = true;
                }
            } while (!validPosition && maxAttempts > 0);

            // Случайный тип еды
            int chance = random.Next(100);

            if (chance < 70)
            {
                Type = FoodType.Normal;
                Points = 10;
                Color = Color.Red;
            }
            else if (chance < 85)
            {
                Type = FoodType.Bonus;
                Points = 50;
                Color = Color.Gold;
            }
            else if (chance < 93)
            {
                Type = FoodType.Speed;
                Points = 20;
                Color = Color.Cyan;
            }
            else if (chance < 98)
            {
                Type = FoodType.Slow;
                Points = 20;
                Color = Color.Magenta;
            }
            else
            {
                Type = FoodType.Golden;
                Points = 100;
                Color = Color.Orange;
            }
        }

        public void Draw(Graphics g, int cellSize)
        {
            Rectangle rect = new Rectangle(
                Position.X * cellSize + 2,
                Position.Y * cellSize + 2,
                cellSize - 4,
                cellSize - 4
            );

            using (SolidBrush brush = new SolidBrush(Color))
            {
                g.FillEllipse(brush, rect);
            }

            // Символ для особой еды
            if (Type != FoodType.Normal)
            {
                string symbol = Type switch
                {
                    FoodType.Bonus => "★",
                    FoodType.Speed => "⚡",
                    FoodType.Slow => "🐢",
                    FoodType.Golden => "👑",
                    _ => ""
                };

                using (Font font = new Font("Arial", 12))
                using (SolidBrush textBrush = new SolidBrush(Color.White))
                {
                    g.DrawString(symbol, font, textBrush, 
                        Position.X * cellSize + 5, 
                        Position.Y * cellSize + 3);
                }
            }
        }
    }
}
